zed: automate Kolla Ansible tag variable definitions

markgoddard opened the pull request:
This change reworks the way that Kolla image tags are defined, making the following improvements possible:

* Automatic promotion of images on merge (as is currently done for package repos)
* New CI job that checks that all specified tags are present in Test Pulp. This shows that they have been built, pushed to Ark, and also synced to Test Pulp on SMS lab
* Automatic definition of the <container>_tag variables for Kolla Ansible

The main visible change is that container image tags are now defined in the kolla_image_tags variable in etc/kayobe/kolla-image-tags.yml. This avoids the previous churn in the etc/kayobe/kolla/globals.yml file.

A new script, tools/kolla-images.py, understands the Kolla image and container relationships, and is able to display the tags used by each image. This output is parsed in the new container promotion workflows added in https://github.com/stackhpc/stackhpc-release-train/pull/193.

markgoddard committed:
Move Kolla image tags to a separate file

This will make it easier to automate various tasks.

jovial commented on etc/kayobe/kolla/globals.yml:
```diff
@@ -10,16 +10,20 @@ enable_docker_repo: "{% raw %}{{ 'overcloud' not in group_names or ansible_facts
+kolla_image_tags:
+{{ kolla_image_tags | to_nice_yaml | indent(width=4, first=true) }}
+
+{% for tag in kolla_image_tags %}
```
```suggestion
{% for tag in kolla_image_tags %}
```

Couldn't we hardcode this list? That way we could get proper precedence:

```
nova_libvirt_tag:  "{% raw %}{{{ kolla_image_tags['nova-libvirt'] | default(nova_tag) | default(openstack_tag) }}{% endraw %}"
```

Then it just leave it permanently in the file. In my opinion it easier to understand what is going on.

jovial replied:
I guess this isn't actually an issue if you had the full list of containers :+1: I was imagining we'd still have groups of containers like 'nova'.

markgoddard replied:
That's an option. I was looking at writing a script that would generate the necessary variables based on kolla_image_tags, but maybe that's trying to be too "clever"?

I'd probably want to put the variables in a separate file and copy them to "declutter" it a bit though.

markgoddard replied:
This version is still using the groups such as nova. The conflict resolution issues in the full image list prevented it from getting finished so I thought I'd try this hybrid approach.

markgoddard committed:
Add base distro version to Kolla image tag definitions

markgoddard committed:
Add kolla-images.py

markgoddard committed:
Remove iscsid from Pulp image list

We no longer require it since Ironic dropped the iscsi deploy driver.

markgoddard committed:
CI: Trigger container image promotion automatically after merge

No more forgotten image promotions!

markgoddard committed:
Automatically define Kolla Ansible tag variables in globals.yml

Use the new kolla-images.py script to automatically define tag variables
in globals.yml based on the tags that are defined in kolla_image_tags.

markgoddard committed:
CI: Add a job to check whether all expected images are in Test Pulp

priteau commented on tools/kolla-images.py:
```diff
@@ -0,0 +1,365 @@
+        if "=" not in line:
+            continue
+        key, value = line.split("=")
+        if key.strip().strip() == "defaultbranch":
```
What is the reason for calling `strip()` twice?

markgoddard replied:
Sleep deprivation? Paranoia? I'll remove it.

priteau commented on tools/kolla-images.py:
```diff
@@ -0,0 +1,365 @@
+            continue
+        key, value = line.split("=")
+        if key.strip().strip() == "defaultbranch":
+            value = value.strip().rstrip()
```
What is the reason for calling both `strip()` and `rstrip()`?

markgoddard replied:
Just one strip should do the job.

priteau commented on tools/kolla-images.py:
```diff
@@ -0,0 +1,365 @@
+        key, value = line.split("=")
+        if key.strip().strip() == "defaultbranch":
+            value = value.strip().rstrip()
+            prefix = "stable/"
+            assert value.startswith(prefix)
+            return value[len(prefix):]
```
Maybe add support for the `unmaintained/` prefix already?

markgoddard replied:
I'll take a look at this.

priteau replied:
It may be required only if:

1) upstream kayobe-config changes `stable/yoga` to `unmaintained/yoga`. Would Yoga be marked eol instead?
2) we merge upstream kayobe-config into stackhpc-kayobe-config (we don't routinely do this at the moment)

priteau commented on tools/kolla-images.py:
```diff
@@ -0,0 +1,365 @@
+    for image, tags in image_tags.items():
+        for _ in range(3):
+            try:
+                output = subprocess.check_output(f"skopeo list-tags docker://{registry}/{namespace}/{image}", shell=True)
```
Is skopeo expected to be present?

markgoddard replied:
Yes. Currently this is just used for local debugging. I could add a `type skopeo` to fail early if it's not installed.

markgoddard replied:
Done

priteau reviewed (changes requested):
Nice code 👍 Just a few minor comments inline. Also please fix merge conflicts.

markgoddard committed:
kolla-images.py: Remove unnecessary strips, support unmaintained branches

markgoddard committed:
kolla-images.py: Check for existence of skopeo

markgoddard committed:
zed: Update bifrost image for Ubuntu Jammy

The previous tag was not in Ark, but this one is present.

markgoddard committed:
kolla-images.py: Fix container to image tag lookup corner case

The failing case was the haproxy image used by the neutron_tls_proxy
container, if a tag is defined for the 'neutron' prefix but not
'haproxy'.

jovial reviewed (approved):
Lets give a whirl. Seems like it will save some time :)
